cpp/src/cluster/detail/single_linkage.cuh

```diff
@@ -47,39 +51,75 @@ namespace cuvs::cluster::agglomerative::detail {
-                      raft::device_matrix_view<value_idx, value_idx> out_dendrogram,
-                      raft::device_vector_view<value_t, value_idx> out_distances,
-                      raft::device_vector_view<value_idx, value_idx> out_sizes)
+template <typename value_t   = float,
+          typename value_idx = int,
+          typename nnz_t     = size_t,
+          typename Accessor  = raft::device_accessor<std::experimental::default_accessor<value_t>>>
+void build_mr_linkage(
+  raft::resources const& handle,
+  raft::mdspan<const value_t, raft::matrix_extent<value_idx>, raft::row_major, Accessor> X,
+  value_idx min_samples,
+  float alpha,
+  cuvs::distance::DistanceType metric,
+  raft::device_vector_view<value_t, value_idx> core_dists,
+  raft::device_coo_matrix_view<value_t, value_idx, value_idx, nnz_t> out_mst,
+  raft::device_matrix_view<value_idx, value_idx> out_dendrogram,
+  raft::device_vector_view<value_t, value_idx> out_distances,
+  raft::device_vector_view<value_idx, value_idx> out_sizes,
+  cuvs::neighbors::all_neighbors::all_neighbors_params all_neighbors_p)
 {
-  size_t m                        = X.extent(0);
-  size_t n                        = X.extent(1);
-  auto mutual_reachability_indptr = raft::make_device_vector<value_idx, value_idx>(handle, m + 1);
-  raft::sparse::COO<value_t, value_idx, nnz_t> mutual_reachability_coo(
-    raft::resource::get_cuda_stream(handle), min_samples * m * 2);
-
-  // Replace this with mutual reachability graph cronstruction from within all_neighbors wrapper.
-  // Reference: https://github.com/rapidsai/cuvs/issues/982
-  cuvs::neighbors::detail::reachability::mutual_reachability_graph<value_idx, value_t, nnz_t>(
-    handle,
-    X.data_handle(),
-    m,
-    n,
-    metric,
-    min_samples,
-    alpha,
-    mutual_reachability_indptr.data_handle(),
-    core_dists.data_handle(),
-    mutual_reachability_coo);
+  size_t m         = X.extent(0);
+  size_t n         = X.extent(1);
+  auto stream      = raft::resource::get_cuda_stream(handle);
+  auto exec_policy = raft::resource::get_thrust_policy(handle);
+
+  auto mr_indptr = raft::make_device_vector<value_idx, value_idx>(handle, m + 1);
+  raft::sparse::COO<value_t, value_idx, nnz_t> mr_coo(stream, min_samples * m * 2);
+
+  auto inds  = raft::make_device_matrix<value_idx, value_idx>(handle, m, min_samples);
+  auto dists = raft::make_device_matrix<value_t, value_idx>(handle, m, min_samples);
+
+  if (all_neighbors_p.metric != metric) {
+    RAFT_LOG_WARN("Setting all neighbors metric to given metrix for build_mr_linkage");
+    all_neighbors_p.metric = metric;
+  }
+  cuvs::neighbors::all_neighbors::build(
+    handle, all_neighbors_p, X, inds.view(), dists.view(), core_dists, alpha);
+
+  // self-loops get max distance
+  rmm::device_uvector<value_idx> coo_rows(min_samples * m, stream);
+  auto coo_rows_counting_itr = thrust::make_counting_iterator<value_idx>(0);
+  thrust::transform(exec_policy,
```

Comment:
`raft::linalg::unaryOp` or similar wherever possible in place of thrust::transform!

Reply:
Agree witht his one- we should always use raft primitives where and when we can.